### PR TITLE
Add handling of custom visual shader nodes from GDExtension

### PR DESCRIPTION
**Pull request by Chaosus**
This will make the editor to intercept them from global class space, so for example (translated example from https://docs.godotengine.org/en/latest/tutorials/plugins/editor/visual_shader_plugins.html) the following C++ class will be handled correctly from its GDExtension module and can be added to the graph as node:

<details>
<summary>visual_shader_node_noise.h</summary>

```cpp
#pragma once

#ifdef WIN32
#include <windows.h>
#endif

#include <godot_cpp/core/binder_common.hpp>

#include <godot_cpp/classes/global_constants.hpp>
#include <godot_cpp/classes/visual_shader_node_custom.hpp>

using namespace godot;

class VisualShaderNodeNoise : public VisualShaderNodeCustom {
GDCLASS(VisualShaderNodeNoise, VisualShaderNodeCustom);

protected:
static void _bind_methods();

public:
virtual String _get_name() const override;
virtual String _get_category() const override;
virtual String _get_description() const override;

virtual int64_t _get_return_icon_type() const override;
virtual int64_t _get_input_port_count() const override;
virtual int64_t _get_input_port_type(int64_t port) const override;
virtual String _get_input_port_name(int64_t port) const override;
virtual int64_t _get_output_port_count() const override;
virtual int64_t _get_output_port_type(int64_t port) const override;
virtual String _get_output_port_name(int64_t port) const override;

virtual String _get_code(const TypedArray<String> &input_vars, const TypedArray<String> &output_vars, Shader::Mode mode, VisualShader::Type type) const override;
virtual String _get_global_code(Shader::Mode mode) const override;

public:
VisualShaderNodeNoise();
~VisualShaderNodeNoise();
};

```

</details>

<details>
<summary>visual_shader_node_noise.cpp</summary>

```cpp
#include "visual_shader_node_noise.h"

#include <godot_cpp/classes/global_constants.hpp>
#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/variant/utility_functions.hpp>

using namespace godot;

////////////////////////////////////////////////////////////////////////////////////////

String VisualShaderNodeNoise::_get_name() const {
return "Noise";
}

String VisualShaderNodeNoise::_get_category() const {
return "MyShaderNodes";
}

String VisualShaderNodeNoise::_get_description() const {
return "Classic Perlin-Noise-3D function (by Curly-Brace)";
}

int64_t VisualShaderNodeNoise::_get_return_icon_type() const {
return PORT_TYPE_SCALAR;
}

int64_t VisualShaderNodeNoise::_get_input_port_count() const {
return 4;
}

int64_t VisualShaderNodeNoise::_get_input_port_type(int64_t port) const {
switch (port) {
case 0:
return PORT_TYPE_VECTOR_3D;
case 1:
return PORT_TYPE_VECTOR_3D;
case 2:
return PORT_TYPE_SCALAR;
case 3:
return PORT_TYPE_SCALAR;
default:
break;
}
return PORT_TYPE_SCALAR;
}

String VisualShaderNodeNoise::_get_input_port_name(int64_t port) const {
switch (port) {
case 0:
return "uv";
case 1:
return "offset";
case 2:
return "scale";
case 3:
return "time";
default:
break;
}
return String();
}

int64_t VisualShaderNodeNoise::_get_output_port_count() const {
return 1;
}

int64_t VisualShaderNodeNoise::_get_output_port_type(int64_t port) const {
return PORT_TYPE_SCALAR;
}

String VisualShaderNodeNoise::_get_output_port_name(int64_t port) const {
return "result";
}

String VisualShaderNodeNoise::_get_code(const TypedArray<String> &input_vars, const TypedArray<String> &output_vars, Shader::Mode mode, VisualShader::Type type) const {
String code;
code = String(output_vars[0]) + String(" = cnoise(vec3((") + String(input_vars[0]) + String(".xy + ") + String(input_vars[1]) + String(".xy) * ") + String(input_vars[2]) + String(", ") + String(input_vars[3]) + String(")) * 0.5 + 0.5;");
return code;
}

String VisualShaderNodeNoise::_get_global_code(Shader::Mode mode) const {
return R"(vec3 mod289_3(vec3 x) {
            return x - floor(x * (1.0 / 289.0)) * 289.0;
        }

        vec4 mod289_4(vec4 x) {
            return x - floor(x * (1.0 / 289.0)) * 289.0;
        }

        vec4 permute(vec4 x) {
            return mod289_4(((x * 34.0) + 1.0) * x);
        }

        vec4 taylorInvSqrt(vec4 r) {
            return 1.79284291400159 - 0.85373472095314 * r;
        }

        vec3 fade(vec3 t) {
            return t * t * t * (t * (t * 6.0 - 15.0) + 10.0);
        }

        // Classic Perlin noise.
        float cnoise(vec3 P) {
            vec3 Pi0 = floor(P); // Integer part for indexing.
            vec3 Pi1 = Pi0 + vec3(1.0); // Integer part + 1.
            Pi0 = mod289_3(Pi0);
            Pi1 = mod289_3(Pi1);
            vec3 Pf0 = fract(P); // Fractional part for interpolation.
            vec3 Pf1 = Pf0 - vec3(1.0); // Fractional part - 1.0.
            vec4 ix = vec4(Pi0.x, Pi1.x, Pi0.x, Pi1.x);
            vec4 iy = vec4(Pi0.yy, Pi1.yy);
            vec4 iz0 = vec4(Pi0.z);
            vec4 iz1 = vec4(Pi1.z);

            vec4 ixy = permute(permute(ix) + iy);
            vec4 ixy0 = permute(ixy + iz0);
            vec4 ixy1 = permute(ixy + iz1);

            vec4 gx0 = ixy0 * (1.0 / 7.0);
            vec4 gy0 = fract(floor(gx0) * (1.0 / 7.0)) - 0.5;
            gx0 = fract(gx0);
            vec4 gz0 = vec4(0.5) - abs(gx0) - abs(gy0);
            vec4 sz0 = step(gz0, vec4(0.0));
            gx0 -= sz0 * (step(0.0, gx0) - 0.5);
            gy0 -= sz0 * (step(0.0, gy0) - 0.5);

            vec4 gx1 = ixy1 * (1.0 / 7.0);
            vec4 gy1 = fract(floor(gx1) * (1.0 / 7.0)) - 0.5;
            gx1 = fract(gx1);
            vec4 gz1 = vec4(0.5) - abs(gx1) - abs(gy1);
            vec4 sz1 = step(gz1, vec4(0.0));
            gx1 -= sz1 * (step(0.0, gx1) - 0.5);
            gy1 -= sz1 * (step(0.0, gy1) - 0.5);

            vec3 g000 = vec3(gx0.x, gy0.x, gz0.x);
            vec3 g100 = vec3(gx0.y, gy0.y, gz0.y);
            vec3 g010 = vec3(gx0.z, gy0.z, gz0.z);
            vec3 g110 = vec3(gx0.w, gy0.w, gz0.w);
            vec3 g001 = vec3(gx1.x, gy1.x, gz1.x);
            vec3 g101 = vec3(gx1.y, gy1.y, gz1.y);
            vec3 g011 = vec3(gx1.z, gy1.z, gz1.z);
            vec3 g111 = vec3(gx1.w, gy1.w, gz1.w);

            vec4 norm0 = taylorInvSqrt(vec4(dot(g000, g000), dot(g010, g010), dot(g100, g100), dot(g110, g110)));
            g000 *= norm0.x;
            g010 *= norm0.y;
            g100 *= norm0.z;
            g110 *= norm0.w;
            vec4 norm1 = taylorInvSqrt(vec4(dot(g001, g001), dot(g011, g011), dot(g101, g101), dot(g111, g111)));
            g001 *= norm1.x;
            g011 *= norm1.y;
            g101 *= norm1.z;
            g111 *= norm1.w;

            float n000 = dot(g000, Pf0);
            float n100 = dot(g100, vec3(Pf1.x, Pf0.yz));
            float n010 = dot(g010, vec3(Pf0.x, Pf1.y, Pf0.z));
            float n110 = dot(g110, vec3(Pf1.xy, Pf0.z));
            float n001 = dot(g001, vec3(Pf0.xy, Pf1.z));
            float n101 = dot(g101, vec3(Pf1.x, Pf0.y, Pf1.z));
            float n011 = dot(g011, vec3(Pf0.x, Pf1.yz));
            float n111 = dot(g111, Pf1);

            vec3 fade_xyz = fade(Pf0);
            vec4 n_z = mix(vec4(n000, n100, n010, n110), vec4(n001, n101, n011, n111), fade_xyz.z);
            vec2 n_yz = mix(n_z.xy, n_z.zw, fade_xyz.y);
            float n_xyz = mix(n_yz.x, n_yz.y, fade_xyz.x);
            return 2.2 * n_xyz;
        })";
}

void VisualShaderNodeNoise::_bind_methods() {
}

VisualShaderNodeNoise::VisualShaderNodeNoise() {
set_input_port_default_value(2, 0.0);
}

VisualShaderNodeNoise::~VisualShaderNodeNoise() {}

```

</details>

The results can be accessed from Add-on menu as usual:
image

Test project (without extension code):
[TestVSGDExt.zip](https://github.com/godotengine/godot/files/10343768/TestVSGDExt.zip)

The extension code (need to be placed to the godot-cpp folder in order to compile):
[test_vs.zip](https://github.com/godotengine/godot/files/10343793/test_vs.zip)

### CODE REVIEW

**Comment by QbieShay** [comment]
Hey @Chaosus , it seems that @paddy-exe attempted to make a GDExtension for visual shaders but isn't working on this PR. The code is here: https://github.com/paddy-exe/GodotVisualShaderExtension/pull/1

Could you share the full extension you made to test this?

**Comment by paddy-exe** [comment]
I still can't compile it but I think this is mostly because I don't have the exact ``godot-cpp`` git commit to checkout. @Chaosus Could you append it please? Otherwise it's hard for me to recreate it.

**Comment by Chaosus** [comment]
@paddy-exe OK, here is the instruction to how to test it:

1. Download: [godot-cpp.zip](https://github.com/godotengine/godot/files/11394431/godot-cpp.zip) - this version is updated to 8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b: 
2. Compile it with the following command: `scons arch=x86_64 target=template_release`.
3. Download the GDExtension project: [test_vs.zip](https://github.com/godotengine/godot/files/11394372/test_vs.zip) and put it to the `godot_cpp` folder (I've updated it a bit so it should work).
4. Goto it and compile with the following command: `scons target=template_release`
5. Compile the Godot with the `de732f0be378b3554aad08158ea695d5b646c893` (the commit from this PR) and run the attached project: [Project.zip](https://github.com/godotengine/godot/files/11394358/Project.zip)
6. (Optional) You can try to change the GDExtension project and recompile it, then put the resulted DLL files (if you on Windows?) to the bin folder of the test project.

Please report of how you've success with it. I think it should work.

**Review by paddy-exe — APPROVED** [review]
image

It works! Awesomemely done!!!

**Comment by clayjohn** [comment]
Would be good to have a review from @touilleMan or @dsnopek as well

**Review by dsnopek — COMMENTED** [review]
Thanks, this is a great addition!

I built from source and tested on Linux. The test project worked, I can see the Noise node in the visual shader graph, and the shader appears to be working! I'm also able to see the new node in the "Add Node..." dialog.

There's only my other comments about some code we might be able to remove.

Great work! :-)

File: editor/plugins/visual_shader_editor_plugin.cpp [hunk 1]
```diff
@@ -1250,49 +1258,59 @@ void VisualShaderEditor::add_custom_type(const String &p_name, const Ref<Script>
...
 	Dictionary dict;
 	dict["script"] = p_custom_node->get_script();
 
-	String name;
-	if (p_custom_node->has_method("_get_name")) {
+	String name = "Unnamed";
+	if (p_use_virtual_calls) {
+		name = p_custom_node->_get_name();
+	} else if (p_custom_node->has_method("_get_name")) {
 		name = (String)p_custom_node->call("_get_name");
```
[the remark this answers is not in the document]
Comment (Chaosus) [reply editor/plugins/visual_shader_editor_plugin.cpp]: Fixed, seems working (with both types of scripts)...

File: editor/plugins/visual_shader_editor_plugin.cpp [hunk 2]
```diff
@@ -1595,29 +1613,60 @@ bool VisualShaderEditor::_is_available(int p_mode) {
...
+				added[key] = dict;
 			}
-			Dictionary dict = get_custom_node_data(ref);
+		}
+	}
+
+	// Add GDExtension classes.
+	{
+		List<StringName> class_list;
+		ClassDB::get_class_list(&class_list);
```
[the remark this answers is not in the document]
Comment (Chaosus) [reply editor/plugins/visual_shader_editor_plugin.cpp:1614]: Nope, they seem to be handled differently, and the GDScript are not appearing in that list. Changing that is beyond of the subject of this PR.

[the remark this answers is not in the document]
Comment (dsnopek) [reply editor/plugins/visual_shader_editor_plugin.cpp:1614]: Ah, ok!

**Review by dsnopek — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by QbieShay** [comment]
Wohooo congratz :partying_face:

**Pull request merged**